src/txmempool.cpp

```diff
@@ -429,5 +438,159 @@ bool CCoinsViewMemPool::HaveCoins(const uint256 &txid) const {
 
 size_t CTxMemPool::DynamicMemoryUsage() const {
     LOCK(cs);
-    return memusage::DynamicUsage(mapTx) + memusage::DynamicUsage(mapNextTx) + memusage::DynamicUsage(mapDeltas) + cachedInnerUsage;
+    // Estimate the overhead of mapTx to be 9 pointers + an allocation, as no exact formula for boost::multi_index_contained is implemented.
+    return memusage::MallocUsage(sizeof(CTxMemPoolEntry) + 9 * sizeof(void*)) * mapTx.size() + memusage::DynamicUsage(mapNextTx) + memusage::DynamicUsage(mapDeltas) + cachedInnerUsage;
+}
+
+size_t CTxMemPool::GuessDynamicMemoryUsage(const CTxMemPoolEntry& entry) const {
+    return memusage::MallocUsage(sizeof(CTxMemPoolEntry) + 9 * sizeof(void*)) + entry.DynamicMemoryUsage() + memusage::IncrementalDynamicUsage(mapNextTx) * entry.GetTx().vin.size();
+}
+
+bool CTxMemPool::StageTrimToSize(size_t sizelimit, const CTxMemPoolEntry& toadd, std::set<uint256>& stage,
+                                 CAmount& nFeesReserved, CAmount& nFeesRemoved) {
+    std::set<uint256> protect;
+    BOOST_FOREACH(const CTxIn& in, toadd.GetTx().vin) {
+        protect.insert(in.prevout.hash);
+    }
+
+    size_t incUsage = GuessDynamicMemoryUsage(toadd);
+    size_t expsize = DynamicMemoryUsage() + incUsage; // Track the expected resulting memory usage of the mempool.
+    if (expsize > sizelimit) {
+        size_t sizeToTrim = std::min(expsize - sizelimit, incUsage);
+        return TrimMempool(sizeToTrim, protect, nFeesReserved, toadd.GetTxSize(), toadd.GetFee(), stage, nFeesRemoved, true);
+    }
+    else
+        return true;
+}
+
+bool CTxMemPool::SurplusTrim(int multiplier, CFeeRate minRelayRate, std::set<uint256>& stage, size_t usageToTrim) {
+    CFeeRate excessRate(multiplier * minRelayRate.GetFeePerK());
+    std::set<uint256> noprotect;
+    CAmount nFeesRemoved = 0;
+    size_t sizeToTrim = usageToTrim / 4;  //Conservatively assume we have transactions at least 1/4 the size of the mempool space they've taken
+    return TrimMempool(usageToTrim, noprotect, 0, sizeToTrim, excessRate.GetFee(sizeToTrim), stage, nFeesRemoved, false);
+}
+
+bool CTxMemPool::TrimMempool(size_t sizeToTrim, std::set<uint256> &protect, CAmount nFeesReserved, size_t sizeToUse, CAmount feeToUse,
+                             std::set<uint256>& stage, CAmount &nFeesRemoved, bool mustTrimAllSize) {
+    size_t usageRemoved = 0;
+    indexed_transaction_set::nth_index<1>::type::reverse_iterator it = mapTx.get<1>().rbegin();
+    int fails = 0; // Number of mempool transactions iterated over that were not included in the stage.
+    int itertotal = 0;
+    int iterextra = mustTrimAllSize ? 10 : 100; //Allow many more iterations to find large size during SurplusTrim
```

Comment:
Do you mind turning these into arguments, that get passed from the 2 callsites instead?